File: app/src/main/java/com/krystianwsul/checkme/gui/tree/TreeNode.java
package com.krystianwsul.checkme.gui.tree;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TreeNode implements Comparable<TreeNode>, NodeContainer {
    private final ModelNode mModelNode;

    private List<TreeNode> mChildTreeNodes;

    @NonNull
    private final NodeContainer mParent;

    private boolean mExpanded;

    private boolean mSelected = false;

    public TreeNode(@NonNull ModelNode modelNode, @NonNull NodeContainer parent, boolean expanded, boolean selected) {
        mModelNode = modelNode;
        mParent = parent;

        mExpanded = expanded;
        mSelected = selected;

        Assert.assertTrue(!mSelected || mModelNode.selectable());
    }

    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder) {
        mModelNode.onBindViewHolder(viewHolder);
    }

    int getItemViewType() {
        return mModelNode.getItemViewType();
    }

    public void update() {
        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(treeNodeCollection.getPosition(this));
    }

    @NonNull
    public ModelNode getModelNode() {
        return mModelNode;
    }

    public void setChildTreeNodes(@NonNull List<TreeNode> childTreeNodes) {
        Assert.assertTrue(!mExpanded || !childTreeNodes.isEmpty());

        mChildTreeNodes = childTreeNodes;

        Collections.sort(mChildTreeNodes);
    }

    public boolean expanded() {
        Assert.assertTrue(!mExpanded || visibleSize() > 1);
        return mExpanded;
    }

    @Override
    public int compareTo(@NonNull TreeNode another) {
        return mModelNode.compareTo(another.mModelNode);
    }

    @NonNull
    public View.OnLongClickListener getOnLongClickListener() {
        return v -> {
            onLongClick();
            return true;
        };
    }


    @NonNull
    public View.OnClickListener getOnClickListener() {
        return v -> {
            if (hasActionMode()) {
                onLongClick();
            } else {
                mModelNode.onClick();
            }
        };
    }

    private void onLongClick() {
        if (!mModelNode.selectable())
            return;

        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        NodeContainer parent = getParent();

        mSelected = !mSelected;

        if (mSelected) {
            incrementSelected();

            if (parent.getSelectedChildren().size() == 1) // first in group
                parent.update();
        } else {
            decrementSelected();

            if (parent.getSelectedChildren().size() == 0) // last in group
                parent.update();
        }

        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(treeNodeCollection.getPosition(this));
    }

    @NonNull
    public NodeContainer getParent() {
        return mParent;
    }

    private boolean hasActionMode() {
        return getTreeViewAdapter().hasActionMode();
    }

    private void incrementSelected() {
        getTreeViewAdapter().incrementSelected();
    }

    private void decrementSelected() {
        getTreeViewAdapter().decrementSelected();
    }

    @NonNull
    private TreeViewAdapter getTreeViewAdapter() {
        return getTreeNodeCollection().mTreeViewAdapter;
    }

    public int displayedSize() {
        if ((!mModelNode.visibleWhenEmpty() && mChildTreeNodes.isEmpty()) || (!mModelNode.visibleDuringActionMode() && hasActionMode())) {
            return 0;
        } else {
            if (mExpanded) {
                return 1 + Stream.of(mChildTreeNodes).map(TreeNode::displayedSize).reduce(0, (a, b) -> a + b);
            } else {
                return 1;
            }
        }
    }

    private int visibleSize() {
        if ((!mModelNode.visibleWhenEmpty() && mChildTreeNodes.isEmpty())) {
            return 0;
        } else {
            if (mExpanded) {
                return 1 + Stream.of(mChildTreeNodes).map(TreeNode::visibleSize).reduce(0, (a, b) -> a + b);
            } else {
                return 1;
            }
        }
    }

    @NonNull
    TreeNode getNode(int position) {
        Assert.assertTrue(position >= 0);
        Assert.assertTrue(!mChildTreeNodes.isEmpty() || mModelNode.visibleWhenEmpty());
        Assert.assertTrue(mModelNode.visibleDuringActionMode() || !hasActionMode());
        Assert.assertTrue(position < displayedSize());

        if (position == 0)
            return this;

        Assert.assertTrue(mExpanded);

        position--;

        for (TreeNode notDoneGroupTreeNode : mChildTreeNodes) {
            if (position < notDoneGroupTreeNode.displayedSize())
                return notDoneGroupTreeNode.getNode(position);

            position -= notDoneGroupTreeNode.displayedSize();
        }

        throw new IndexOutOfBoundsException();
    }

    public int getPosition(@NonNull TreeNode treeNode) {
        if (treeNode == this)
            return 0;

        if (!mExpanded)
            return -1;

        int offset = 1;
        for (TreeNode childTreeNode : mChildTreeNodes) {
            int position = childTreeNode.getPosition(treeNode);
            if (position >= 0)
                return offset + position;
            offset += childTreeNode.displayedSize();
        }

        return -1;
    }

    public boolean isSelected() {
        Assert.assertTrue(!mSelected || mModelNode.selectable());

        return mSelected;
    }

    public void unselect() {
        Assert.assertTrue(!mSelected || mModelNode.selectable());

        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        if (mSelected) {
            Assert.assertTrue(mModelNode.selectable());

            mSelected = false;
            treeNodeCollection.mTreeViewAdapter.notifyItemChanged(treeNodeCollection.getPosition(this));
        }

        List<TreeNode> selected = getSelectedNodes()
                .collect(Collectors.toList());

        if (!selected.isEmpty()) {
            Assert.assertTrue(mExpanded);

            Stream.of(selected)
                    .forEach(TreeNode::unselect);

            treeNodeCollection.mTreeViewAdapter.notifyItemChanged(treeNodeCollection.getPosition(this));
        }
    }

    public void selectAll() {
        Assert.assertTrue(!mSelected);

        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        if (mModelNode.selectable()) {
            mSelected = true;

            treeNodeCollection.mTreeViewAdapter.notifyItemChanged(treeNodeCollection.getPosition(this));

            treeNodeCollection.mTreeViewAdapter.incrementSelected();
        }

        if (mExpanded) {
            Assert.assertTrue(!mChildTreeNodes.isEmpty());

            Stream.of(mChildTreeNodes).forEach(TreeNode::selectAll);
        }
    }

    @NonNull
    Stream<TreeNode> getSelectedNodes() {
        Assert.assertTrue(mChildTreeNodes != null);

        Assert.assertTrue(!mSelected || mModelNode.selectable());

        ArrayList<TreeNode> selectedTreeNodes = new ArrayList<>();

        if (mSelected)
            selectedTreeNodes.add(this);

        selectedTreeNodes.addAll(Stream.of(mChildTreeNodes)
                .flatMap(TreeNode::getSelectedNodes)
                .collect(Collectors.toList()));

        return Stream.of(selectedTreeNodes);
    }

    @NonNull
    public View.OnClickListener getExpandListener() {
        return v -> {
            Assert.assertTrue(!mChildTreeNodes.isEmpty());

            TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

            Assert.assertTrue(!(!getSelectedChildren().isEmpty() && hasActionMode()));

            int position = treeNodeCollection.getPosition(this);
            Assert.assertTrue(position >= 0);

            if (mExpanded) { // hiding
                Assert.assertTrue(getSelectedChildren().isEmpty());

                int displayedSize = displayedSize();
                mExpanded = false;
                treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(position + 1, displayedSize - 1);
            } else { // showing
                mExpanded = true;
                treeNodeCollection.mTreeViewAdapter.notifyItemRangeInserted(position + 1, displayedSize() - 1);
            }

            if (position > 0) {
                treeNodeCollection.mTreeViewAdapter.notifyItemRangeChanged(position - 1, 2);
            } else {
                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(position);
            }
        };
    }

    private boolean visible() {
        NodeContainer nodeContainer = getParent();

        if (!mModelNode.visibleDuringActionMode() && hasActionMode())
            return false;

        if (!mModelNode.visibleWhenEmpty() && mChildTreeNodes.isEmpty())
            return false;

        if (nodeContainer instanceof TreeNodeCollection) {
            return true;
        } else {
            Assert.assertTrue(nodeContainer instanceof TreeNode);
            TreeNode parent = (TreeNode) nodeContainer;

            return (parent.visible() && parent.expanded());
        }
    }

    @Override
    public void remove(@NonNull TreeNode childTreeNode) {
        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        Assert.assertTrue(!mChildTreeNodes.isEmpty());
        Assert.assertTrue(mChildTreeNodes.contains(childTreeNode));

        int childDisplayedSize = childTreeNode.displayedSize();

        int oldParentPosition = treeNodeCollection.getPosition(this);

        int oldChildPosition = treeNodeCollection.getPosition(childTreeNode);

        boolean expanded = expanded();
        boolean visible = visible();

        mChildTreeNodes.remove(childTreeNode);

        if (!visible)
            return;

        Assert.assertTrue(oldParentPosition >= 0);

        if (expanded) {
            Assert.assertTrue(oldChildPosition >= 0);

            if (Stream.of(mChildTreeNodes).map(TreeNode::displayedSize).reduce(0, (lhs, rhs) -> lhs + rhs) == 0) {
                mExpanded = false;

                if (oldParentPosition == 0) {
                    if (mModelNode.visibleWhenEmpty()) {
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                        treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(oldParentPosition + 1, childDisplayedSize);
                    } else {
                        treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(oldParentPosition, 1 + childDisplayedSize);
                    }
                } else {
                    if (mModelNode.visibleWhenEmpty()) {
                        treeNodeCollection.mTreeViewAdapter.notifyItemRangeChanged(oldParentPosition - 1, 2);
                        treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(oldParentPosition + 1, childDisplayedSize);
                    } else {
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
                        treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(oldParentPosition, 1 + childDisplayedSize);
                    }
                }
            } else {
                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(oldChildPosition, childDisplayedSize);

                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldChildPosition - 1);

                if (oldParentPosition > 0)
                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
            }
        } else {
            if (Stream.of(mChildTreeNodes).map(TreeNode::displayedSize).reduce(0, (lhs, rhs) -> lhs + rhs) == 0) {
                if (oldParentPosition == 0) {
                    if (mModelNode.visibleWhenEmpty()) {
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                    } else {
                        treeNodeCollection.mTreeViewAdapter.notifyItemRemoved(oldParentPosition);
                    }
                } else {
                    if (mModelNode.visibleWhenEmpty()) {
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                    } else {
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
                        treeNodeCollection.mTreeViewAdapter.notifyItemRemoved(oldParentPosition);
                    }
                }
            } else {
                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);

                if (oldParentPosition > 0)
                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
            }
        }
    }

    public void removeAll() {
        List<TreeNode> oldChildTreeNodes = new ArrayList<>(mChildTreeNodes);
        Stream.of(oldChildTreeNodes)
                .forEach(this::remove);
    }

    @Override
    public void add(@NonNull TreeNode childTreeNode) {
        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        if (mExpanded) {
            if (mModelNode.visibleWhenEmpty()) {
                int oldParentPosition = treeNodeCollection.getPosition(this);
                Assert.assertTrue(oldParentPosition >= 0);

                mChildTreeNodes.add(childTreeNode);

                Collections.sort(mChildTreeNodes);

                int newChildPosition = treeNodeCollection.getPosition(childTreeNode);
                Assert.assertTrue(newChildPosition >= 0);

                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                treeNodeCollection.mTreeViewAdapter.notifyItemInserted(newChildPosition);

                boolean last = (oldParentPosition + displayedSize() - 1 == newChildPosition);

                if (last)
                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(newChildPosition - 1);

                if (oldParentPosition > 0)
                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
            } else {
                if (mChildTreeNodes.isEmpty()) {
                    mChildTreeNodes.add(childTreeNode);

                    Collections.sort(mChildTreeNodes);

                    int newParentPosition = treeNodeCollection.getPosition(this);

                    treeNodeCollection.mTreeViewAdapter.notifyItemInserted(newParentPosition + 1);

                    if (newParentPosition > 0)
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(newParentPosition - 1);
                } else {
                    int oldParentPosition = treeNodeCollection.getPosition(this);
                    Assert.assertTrue(oldParentPosition >= 0);

                    mChildTreeNodes.add(childTreeNode);

                    Collections.sort(mChildTreeNodes);

                    int newChildPosition = treeNodeCollection.getPosition(childTreeNode);
                    Assert.assertTrue(newChildPosition >= 0);

                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition);
                    treeNodeCollection.mTreeViewAdapter.notifyItemInserted(newChildPosition);

                    boolean last = (oldParentPosition + displayedSize() - 1 == newChildPosition);

                    if (last)
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(newChildPosition - 1);

                    if (oldParentPosition > 0)
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(oldParentPosition - 1);
                }
            }
        } else {
            mChildTreeNodes.add(childTreeNode);

            Collections.sort(mChildTreeNodes);

            if (mModelNode.visibleDuringActionMode() || !hasActionMode()) {
                int newParentPosition = treeNodeCollection.getPosition(this);
                Assert.assertTrue(newParentPosition >= 0);

                if (!mModelNode.visibleWhenEmpty() && mChildTreeNodes.size() == 1) {
                    treeNodeCollection.mTreeViewAdapter.notifyItemInserted(newParentPosition);

                    if (newParentPosition > 0)
                        treeNodeCollection.mTreeViewAdapter.notifyItemChanged(newParentPosition - 1);
                } else {
                    treeNodeCollection.mTreeViewAdapter.notifyItemChanged(newParentPosition);
                }
            }
        }
    }

    public boolean getSeparatorVisibility() {
        NodeContainer parent = getParent();

        Assert.assertTrue(parent.expanded());

        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        int positionInCollection = treeNodeCollection.getPosition(this);
        Assert.assertTrue(positionInCollection >= 0);

        if (positionInCollection == treeNodeCollection.displayedSize() - 1)
            return false;

        if (parent.getPosition(this) == parent.displayedSize() - 1)
            return true;

        TreeNode nextTreeNode = treeNodeCollection.getNode(positionInCollection + 1);

        return (nextTreeNode.expanded() || mModelNode.separatorVisibleWhenNotExpanded());
    }

    @NonNull
    @Override
    public List<TreeNode> getSelectedChildren() {
        Assert.assertTrue(!mChildTreeNodes.isEmpty());

        return Stream.of(mChildTreeNodes)
                .filter(TreeNode::isSelected)
                .collect(Collectors.toList());
    }

    void onCreateActionMode() {
        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        int position = treeNodeCollection.getPosition(this);
        Assert.assertTrue(position >= 0);

        if (mModelNode.visibleDuringActionMode()) {
            treeNodeCollection.mTreeViewAdapter.notifyItemChanged(position);

            if (mExpanded)
                Stream.of(mChildTreeNodes)
                        .forEach(TreeNode::onCreateActionMode);
        } else {
            if (mChildTreeNodes.size() > 0) {
                if (mExpanded) {
                    treeNodeCollection.mTreeViewAdapter.notifyItemRangeRemoved(position, visibleSize());
                } else {
                    treeNodeCollection.mTreeViewAdapter.notifyItemRemoved(position);
                }
            } else if (mModelNode.visibleWhenEmpty()) {
                treeNodeCollection.mTreeViewAdapter.notifyItemRemoved(position);
            }

            if (position > 0)
                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(position - 1);
        }
    }

    void onDestroyActionMode() {
        TreeNodeCollection treeNodeCollection = getTreeNodeCollection();

        int position = treeNodeCollection.getPosition(this);
        Assert.assertTrue(position >= 0);

        if (mModelNode.visibleDuringActionMode()) {
            treeNodeCollection.mTreeViewAdapter.notifyItemChanged(position);

            if (mExpanded)
                Stream.of(mChildTreeNodes)
                        .forEach(TreeNode::onDestroyActionMode);
        } else {
            if (mChildTreeNodes.size() > 0) {
                if (mExpanded) {
                    treeNodeCollection.mTreeViewAdapter.notifyItemRangeInserted(position, displayedSize());
                } else {
                    treeNodeCollection.mTreeViewAdapter.notifyItemInserted(position);
                }
            } else if (mModelNode.visibleWhenEmpty()) {
                treeNodeCollection.mTreeViewAdapter.notifyItemInserted(position);
            }

            if (position > 0)
                treeNodeCollection.mTreeViewAdapter.notifyItemChanged(position - 1);
        }
    }

    @NonNull
    @Override
    public TreeNodeCollection getTreeNodeCollection() {
        return getParent().getTreeNodeCollection();
    }

    @NonNull
    public List<TreeNode> getAllChildren() {
        return mChildTreeNodes;
    }

    @Override
    public int getIndentation() {
        return getParent().getIndentation() + 1;
    }

    public void select() {
        Assert.assertTrue(!mSelected);
        Assert.assertTrue(mModelNode.selectable());

        onLongClick();
    }

    public void deselect() {
        Assert.assertTrue(mSelected);

        onLongClick();
    }
}

File: treeadapter/src/main/java/com/krystianwsul/treeadapter/TreeNode.kt
package com.krystianwsul.treeadapter

import android.os.Parcelable
import androidx.recyclerview.widget.RecyclerView
import kotlinx.parcelize.Parcelize

class TreeNode<T : TreeHolder>(
    val modelNode: ModelNode<T>,
    var parent: NodeContainer<T>,
    private var selected: Boolean = false,
    private val initialExpansionState: ExpansionState? = null,
) : Comparable<TreeNode<T>>, NodeContainer<T> {

    override val treeNodeCollection by lazy { parent.treeNodeCollection }

    private lateinit var expansionState: ExpansionState

    fun getSaveExpansionState() = expansionState.takeIf { !it.isDefault }

    override val isExpanded get() = expansionState.isExpanded

    private lateinit var childTreeNodes: MutableList<TreeNode<T>>

    val itemViewType = modelNode.itemViewType

    fun onLongClick(viewHolder: RecyclerView.ViewHolder) {
        checkChildTreeNodesSet()

        if (!modelNode.isSelectable) return

        if (modelNode.tryStartDrag(viewHolder)) {
            modelNode.onBindViewHolder(viewHolder, true)
        } else {
            if (!selected) modelNode.forceSelected(viewHolder)

            treeViewAdapter.updateDisplayedNodes(this::toggleSelected)
        }
    }

    fun onClick(holder: T) {
        if (hasActionMode()) {
            if (!selected) modelNode.forceSelected(holder)

            treeViewAdapter.updateDisplayedNodes(this::toggleSelected)
        } else {
            modelNode.onClick(holder)
        }
    }

    val treeViewAdapter by lazy { treeNodeCollection.treeViewAdapter }

    val isSelected: Boolean
        get() {
            check(!selected || modelNode.isSelectable)

            return selected
        }

    val selectedNodes: List<TreeNode<T>>
        get() {
            checkChildTreeNodesSet()

            val selectedTreeNodes = ArrayList<TreeNode<T>>()

            if (selected) {
                check(modelNode.isSelectable)
                check(modelNodeVisible)

                selectedTreeNodes += this
            }

            selectedTreeNodes += childTreeNodes.flatMap { it.selectedNodes }

            return selectedTreeNodes
        }

    val expandVisible: Boolean
        get() {
            checkChildTreeNodesSet()

            if (!visible()) throw InvisibleNodeException()

            if (childTreeNodes.none { it.canBeShown() }) return false

            return true
        }

    val state get() = State(isExpanded, isSelected, expandVisible, separatorVisible, modelNode.state)

    // hiding
    // showing
    fun onExpandClick() {
        checkChildTreeNodesSet()

        if (childTreeNodes.isEmpty()) throw EmptyExpandedException()

        treeViewAdapter.updateDisplayedNodes { placeholder ->
            expansionState.user = if (isExpanded) { // collapsing
                childTreeNodes.forEach { it.setSelected(false, placeholder, false, true) }

                false
            } else { // expanding
                if (selected)
                    childTreeNodes.forEach { it.setSelected(true, placeholder, false, true) }

                true
            }
        }
    }

    val separatorVisible: Boolean
        get() {
            if (!parent.isExpanded) throw InvisibleNodeException()

            val positionInCollection = treeNodeCollection.getPosition(this, PositionMode.DISPLAYED)
            check(positionInCollection >= 0)

            if (positionInCollection == treeNodeCollection.displayedNodes.size - 1) return false

            if (parent.getPosition(this, PositionMode.DISPLAYED) == parent.displayedNodes.size - 1) {
                if (modelNode.inheritParentBottomSeparator && parent.showInheritableBottomSeparator) return true

                if (parent.wantsSeparators) return true
            }

            val nextTreeNode = treeNodeCollection.getNode(positionInCollection + 1, PositionMode.DISPLAYED)
            return modelNode.isSeparatorVisibleWhenNotExpanded || nextTreeNode.wantsSeparators
        }

    override val wantsSeparators get() = displayedChildNodes.any { it.modelNode.showSeparatorWhenParentExpanded }

    val allChildren: List<TreeNode<T>> get() = childTreeNodes

    override val showInheritableBottomSeparator: Boolean
        get() {
            if (this == parent.displayedDirectChildNodes.last()) return parent.wantsSeparators

            return false
        }

    init {
        if (selected && !modelNode.isSelectable) throw NotSelectableSelectedException()
    }

    fun setChildTreeNodes(childTreeNodes: List<TreeNode<T>>) {
        // todo add delegate with final initialized state, move majority of function calls into it
        if (this::childTreeNodes.isInitialized) throw SetChildTreeNodesCalledTwiceException()

        expansionState = initialExpansionState?.takeIf { childTreeNodes.isNotEmpty() } ?: ExpansionState()

        this.childTreeNodes = childTreeNodes.sorted().toMutableList()
    }

    fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder) = modelNode.onBindViewHolder(viewHolder)

    fun onPayload(viewHolder: RecyclerView.ViewHolder) = modelNode.onPayload(viewHolder)

    override fun compareTo(other: TreeNode<T>) = modelNode.compareTo(other.modelNode)

    private fun toggleSelected(placeholder: TreeViewAdapter.Placeholder) =
        setSelected(!selected, placeholder, true, true)

    private fun setSelected(
        selected: Boolean,
        placeholder: TreeViewAdapter.Placeholder,
        recurseParent: Boolean,
        recurseChildren: Boolean,
    ) {
        checkChildTreeNodesSet()

        if (!modelNode.isSelectable) return
        if (this.selected == selected) return

        this.selected = selected

        if (selected)
            incrementSelected(placeholder)
        else
            decrementSelected(placeholder)

        if (recurseChildren && isExpanded && modelNode.propagateSelection) {
            childTreeNodes.forEach { it.setSelected(selected, placeholder, false, true) }
        }

        if (recurseParent && !selected) {
            (parent as? TreeNode)?.takeIf {
                it.modelNode.propagateSelection
            }?.setSelected(false, placeholder, true, false)
        }
    }

    private fun hasActionMode() = treeViewAdapter.hasActionMode

    private fun incrementSelected(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.incrementSelected(placeholder)

    private fun decrementSelected(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.decrementSelected(placeholder)

    override fun getNode(position: Int, positionMode: PositionMode): TreeNode<T> {
        checkChildTreeNodesSet()

        check(position in positionMode.getRecursiveNodes(this).indices)

        if (position == 0) return this

        check(isExpanded)

        var currPosition = position - 1

        for (notDoneGroupTreeNode in positionMode.getDirectChildNodes(this)) {
            val notDoneDisplayedSize = positionMode.getRecursiveNodes(notDoneGroupTreeNode).size

            if (currPosition < notDoneDisplayedSize) return notDoneGroupTreeNode.getNode(currPosition, positionMode)

            currPosition -= notDoneDisplayedSize
        }

        throw IndexOutOfBoundsException()
    }

    override fun getPosition(positionMode: PositionMode, matcher: (TreeNode<T>) -> Boolean): Int {
        checkChildTreeNodesSet()

        if (matcher(this)) return 0
        if (!isExpanded) return -1

        var offset = 1
        for (childTreeNode in positionMode.getDirectChildNodes(this)) {
            val position = childTreeNode.getPosition(positionMode, matcher)

            if (position >= 0) return offset + position

            offset += positionMode.getRecursiveNodes(childTreeNode).size
        }

        return -1
    }

    fun unselect(placeholder: TreeViewAdapter.Placeholder) {
        check(!selected || modelNode.isSelectable)

        if (selected) {
            check(modelNode.isSelectable)
            check(modelNodeVisible)

            selected = false
        }

        val selected = selectedNodes

        if (selected.isNotEmpty()) {
            check(isExpanded)

            selected.forEach { it.unselect(placeholder) }
        }
    }

    fun select(placeholder: TreeViewAdapter.Placeholder) {
        checkChildTreeNodesSet()

        if (!modelNode.isSelectable) return

        check(modelNodeVisible)

        selected = true

        treeViewAdapter.incrementSelected(placeholder)
    }

    override val displayedNodes: List<TreeNode<T>>
        get() {
            val locker = getLocker()
            locker?.displayedNodes?.let { return it }

            val result = if (canBeShown()) listOf(this) + displayedChildNodes else listOf()

            locker?.displayedNodes = result
            return result
        }

    override val displayableNodes: List<TreeNode<T>>
        get() {
            if (!canBeShown()) return listOf()

            return listOf(this) + childTreeNodes.flatMap { it.displayableNodes }
        }

    override val displayedChildNodes: List<TreeNode<T>>
        get() {
            check(canBeShown())

            return if (isExpanded)
                childTreeNodes.flatMap { it.displayedNodes }
            else
                listOf()
        }

    override val displayedDirectChildNodes: List<TreeNode<T>>
        get() {
            check(visible())

            return childTreeNodes.filter { it.visible() }
        }

    private fun getLocker() = treeViewAdapter.locker?.getNodeLocker(this)

    private val modelNodeVisible get() = modelNode.isVisible(hasActionMode(), childTreeNodes.any { it.canBeShown() })

    /**
     * todo: consider adding a cache that can be used when these values are known not to change, such as after
     * updateDisplayedNodes finishes running
     */
    fun canBeShown(): Boolean {
        checkChildTreeNodesSet()

        return modelNodeVisible
    }

    private val matchesSearch = modelNode.matchesSearch

    private val childHierarchyMatchesSearch: Boolean get() = childTreeNodes.any { it.matchesSearch || it.childHierarchyMatchesSearch }

    fun visible(): Boolean {
        checkChildTreeNodesSet()

        if (!canBeShown()) return false

        return if (parent is TreeNodeCollection<T>) {
            true
        } else {
            (parent as TreeNode).visible() && parent.isExpanded
        }
    }

    override fun remove(treeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        checkChildTreeNodesSet()

        if (!childTreeNodes.contains(treeNode))
            throw NoSuchNodeException()

        val visible = visible()

        childTreeNodes.remove(treeNode)

        if (!visible) return

        if (isExpanded && 0 == childTreeNodes.sumOf { it.displayedNodes.size }) expansionState = ExpansionState()
    }

    override fun add(treeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        checkChildTreeNodesSet()

        childTreeNodes.add(treeNode)
        childTreeNodes.sort()
    }

    override val selectedChildren: List<TreeNode<T>>
        get() {
            checkChildTreeNodesSet()

            if (childTreeNodes.isEmpty()) throw NoChildrenException()

            return childTreeNodes.filter { it.isSelected }
        }

    override val indentation by lazy { parent.indentation + 1 }

    fun resetExpansion(onlyProgrammatic: Boolean, placeholder: TreeViewAdapter.Placeholder) {
        childTreeNodes.forEach { it.resetExpansion(onlyProgrammatic, placeholder) }

        if (onlyProgrammatic)
            expansionState.programmatic = false
        else
            expansionState = ExpansionState()
    }

    fun expandMatching() {
        checkChildTreeNodesSet()

        if (childHierarchyMatchesSearch) expansionState.programmatic = true

        childTreeNodes.forEach { it.expandMatching() }
    }

    private fun checkChildTreeNodesSet() {
        if (!this::childTreeNodes.isInitialized) throw SetChildTreeNodesNotCalledException()
    }

    override fun swapNodePositions(
        fromTreeNode: TreeNode<T>,
        toTreeNode: TreeNode<T>,
        placeholder: TreeViewAdapter.Placeholder,
    ) {
        check(treeViewAdapter.locker == null)

        val fromPosition = childTreeNodes.indexOf(fromTreeNode)
        check(fromPosition >= 0)

        val toPosition = childTreeNodes.indexOf(toTreeNode)
        check(toPosition >= 0)

        childTreeNodes.apply {
            removeAt(fromPosition)
            add(toPosition, fromTreeNode)
        }
    }

    override fun removeForSwap(fromTreeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        check(treeViewAdapter.locker == null)

        val fromPosition = childTreeNodes.indexOf(fromTreeNode)
        check(fromPosition >= 0)

        childTreeNodes.removeAt(fromPosition)
    }

    override fun addForSwap(fromTreeNode: TreeNode<T>, toTreeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        check(treeViewAdapter.locker == null)

        val toPosition = childTreeNodes.indexOf(toTreeNode)
        check(toPosition >= 0)

        fromTreeNode.parent = this
        childTreeNodes.add(toPosition, fromTreeNode)
    }

    class SetChildTreeNodesNotCalledException : InitializationException("TreeNode.setChildTreeNodes() has not been called.")

    class SetChildTreeNodesCalledTwiceException :
        InitializationException("TreeNode.setChildTreeNodes() has already been called.")

    class NotSelectableSelectedException :
        IllegalStateException("A TreeNode cannot be selected if its ModelNode is not selectable.")

    class EmptyExpandedException : IllegalStateException("A TreeNode cannot be expanded if it has no children.")

    class NoSuchNodeException : IllegalArgumentException("The given node is not a direct descendant of this TreeNode.")

    class InvisibleNodeException : UnsupportedOperationException("This operation is meaningless if the node is not visible.")

    class NoChildrenException : UnsupportedOperationException("Can't get selected children of a node that has no children.")

    data class State(
        val isExpanded: Boolean,
        val isSelected: Boolean,
        val expandVisible: Boolean,
        val separatorVisibility: Boolean,
        val modelState: ModelState,
    ) {

        fun getPayload(other: State) = if (isExpanded == other.isExpanded &&
            isSelected == other.isSelected &&
            expandVisible == other.expandVisible &&
            separatorVisibility != other.separatorVisibility &&
            modelState == other.modelState
        ) PayloadSeparator else null
    }

    object PayloadSeparator

    override val id get() = modelNode.id

    @Parcelize
    data class ExpansionState(var programmatic: Boolean = false, var user: Boolean? = null) : Parcelable {

        companion object {

            private val default = ExpansionState()
        }

        val isExpanded get() = user ?: programmatic

        val isDefault get() = this == default
    }
}
